# HG changeset patch
# User Lachlan Andrew <lachlanbis@gmail.com>
# Date 1464430136 -36000
#      Sat May 28 20:08:56 2016 +1000
# Node ID 603f0d752de2efa191d67d614f66aca17e50355d
# Parent  30a0756e6318bce309fe0d5be38de70b20d73351
Wrap search for "next bookmark" and "prev bookmark" (bug #48044)

* file-editor-tab.cc (file_editor_tab::next_bookmark,
file_editor_tab::previous_bookmark):
Retry from start/end if next/previous was -1.

* file-editor.cc (file_editor::construct):
Put "Next Bookmark" before "Previous Bookmark" in the menu.

diff --git a/libgui/src/m-editor/file-editor-tab.cc b/libgui/src/m-editor/file-editor-tab.cc
--- a/libgui/src/m-editor/file-editor-tab.cc
+++ b/libgui/src/m-editor/file-editor-tab.cc
@@ -897,16 +897,20 @@ file_editor_tab::next_bookmark (const QW
 
   int line, cur;
   _edit_area->getCursorPosition (&line, &cur);
 
   line++; // Find bookmark strictly after the current line.
 
   int nextline = _edit_area->markerFindNext (line, (1 << marker::bookmark));
 
+  // Wrap.
+  if (nextline == -1)
+    nextline = _edit_area->markerFindNext (1, (1 << marker::bookmark));
+
   _edit_area->setCursorPosition (nextline, 0);
 }
 
 // Move the text cursor to the closest bookmark
 // before the current line.
 void
 file_editor_tab::previous_bookmark (const QWidget *ID)
 {
@@ -915,16 +919,21 @@ file_editor_tab::previous_bookmark (cons
 
   int line, cur;
   _edit_area->getCursorPosition (&line, &cur);
 
   line--; // Find bookmark strictly before the current line.
 
   int prevline = _edit_area->markerFindPrevious (line, (1 << marker::bookmark));
 
+  // Wrap.  Should use the last line of the file, not 1<<15
+  if (prevline == -1)
+    prevline = _edit_area->markerFindPrevious (_edit_area->lines (),
+                                               (1 << marker::bookmark));
+
   _edit_area->setCursorPosition (prevline, 0);
 }
 
 void
 file_editor_tab::remove_bookmark (const QWidget *ID)
 {
   if (ID != this)
     return;
diff --git a/libgui/src/m-editor/file-editor.cc b/libgui/src/m-editor/file-editor.cc
--- a/libgui/src/m-editor/file-editor.cc
+++ b/libgui/src/m-editor/file-editor.cc
@@ -1710,20 +1710,20 @@ file_editor::construct (void)
 
   _move_to_matching_brace = add_action (_edit_nav_menu, QIcon (),
           tr ("Move to Matching Brace"), SLOT (request_move_match_brace (bool)));
   _sel_to_matching_brace = add_action (_edit_nav_menu, QIcon (),
           tr ("Select to Matching Brace"), SLOT (request_sel_match_brace (bool)));
 
   _edit_nav_menu->addSeparator ();
 
+  _next_bookmark_action = add_action (_edit_nav_menu, QIcon (),
+          tr ("&Next Bookmark"), SLOT (request_next_bookmark (bool)));
   _previous_bookmark_action = add_action (_edit_nav_menu, QIcon (),
           tr ("Pre&vious Bookmark"), SLOT (request_previous_bookmark (bool)));
-  _next_bookmark_action = add_action (_edit_nav_menu, QIcon (),
-          tr ("&Next Bookmark"), SLOT (request_next_bookmark (bool)));
   _toggle_bookmark_action = add_action (_edit_nav_menu, QIcon (),
           tr ("Toggle &Bookmark"), SLOT (request_toggle_bookmark (bool)));
   _remove_bookmark_action = add_action (_edit_nav_menu, QIcon (),
           tr ("&Remove All Bookmarks"), SLOT (request_remove_bookmark (bool)));
 
   _edit_menu->addSeparator ();
 
   _preferences_action = add_action (_edit_menu, resource_manager::icon ("preferences-system"),
